src/mlpack/methods/fastmks/fastmks_rules_impl.hpp

```diff
@@ -46,6 +45,40 @@ FastMKSRules<KernelType, TreeType>::FastMKSRules(
   // dereference null pointers.
   traversalInfo.LastQueryNode() = (TreeType*) this;
   traversalInfo.LastReferenceNode() = (TreeType*) this;
+
+  // Let's build the list of candidate points for each query point.
+  // It will be initialized with k candidates: (-DBL_MAX, size_t() - 1)
+  // The list of candidates will be updated when visiting new points with the
+  // BaseCase() method.
+  const Candidate def = std::make_pair(-DBL_MAX, size_t() - 1);
+
+  CandidateList cList(k, def);
+  std::vector<CandidateList> tmp(querySet.n_cols, cList);
+  candidates.swap(tmp);
+}
+
+template<typename KernelType, typename TreeType>
+void FastMKSRules<KernelType, TreeType>::GetResults(
+    arma::Mat<size_t>& indices,
+    arma::mat& products)
+{
+  indices.set_size(k, querySet.n_cols);
+  products.set_size(k, querySet.n_cols);
+
+  for (size_t i = 0; i < querySet.n_cols; i++)
+  {
+    CandidateList& pqueue = candidates[i];
+    typedef typename CandidateList::iterator Iterator;
+
+    for (Iterator end = pqueue.end(); end != pqueue.begin(); --end)
+      std::pop_heap(pqueue.begin(), end, CandidateCmp());
```

Comment:
Would it be better to use `sort_heap()` here, since we don't need the heap properties anymore?